lib/linux_admin/ssh_agent.rb

```diff
@@ -0,0 +1,113 @@
+require 'net/ssh'
+require 'fileutils'
+module LinuxAdmin
+  class SshAgent
+    attr_reader :pid
+    attr_reader :sock
+
+    def initialize(ssh_private_key, options = {})
+      @sock         = options[:agent_socket] if options[:agent_socket]
+      @private_key = ssh_private_key
+    end
+
+    def start
+      if @sock
+        FileUtils.mkdir_p(File.dirname(@sock))
+        agent_details = `ssh-agent -a #{@sock}`
+      else
+        agent_details = `ssh-agent`
+        @sock         = agent_details.split("=")[1].split(" ")[0].chop
+      end
+      @pid          = agent_details.split("=")[2].split(" ")[0].chop
+      IO.popen({"SSH_AUTH_SOCK" => @sock, "SSH_AGENT_PID" => @pid}, ["ssh-add", "-"], :mode => 'w') do |f|
+        f.puts(@private_key)
+      end
+      if $?.to_i != 0
+        raise StandardError, "Couldn't add key to agent"
+      end
+    end
+
+    def stop
+      if process_exists?(@pid)
+        system({"SSH_AGENT_PID" => @pid}, "(ssh-agent -k) &> /dev/null")
+        if File.exist?(@sock)
+          File.delete(@sock)
+        end
+      end
+    end
+
+    def perform_commands(ip, username, commands)
+      result = nil
+      Net::SSH.start(ip, username, :paranoid => false, :forward_agent => true, :agent_socket_factory => -> { UNIXSocket.open(@sock) }) do |ssh|
+        if block_given?
+          yield ssh
+        else
+          commands.each do |cmd|
+            result = ssh_exec!(ssh, cmd)
+            result[:last_command] = cmd
+            break if result[:exit_status] != 0
+          end
+        end
+      end
+      result
+    end
+
+    def check_connection(ip, username, sub_ips, additional_flags)
```

Comment:
removing this method, i can implement it using a block to perform_commands method.